zerver/lib/message.py

```diff
@@ -614,17 +614,19 @@ def get_user_ids_having_usermessage_row_for_messages(message_ids: list[int]) ->
             usermessages = usermessages.exclude(user_profile__long_term_idle=True)
         return set(usermessages.values_list("user_profile_id", flat=True))
 
-    sample_message = messages[0]
-    if not sample_message.is_channel_message:
+    if not is_channel_message:
         # For DM, event is sent to users who actually received the message.
         return get_user_ids_having_usermessage_row_for_messages(message_ids)
 
-    channel_id = sample_message.recipient.type_id
     if channel is None:
-        channel = Stream.objects.get(id=channel_id)
+        sample_message_id = message_ids[0]
+        recipient_id = Message.objects.filter(id=sample_message_id).values_list(
+            "recipient_id", flat=True
+        )[0]
+        channel = Stream.objects.get(recipient_id=recipient_id)
```

Comment:
I think you can cut a query here; if you fetch `channel`, `recipient_id` can be computed from that, without doing two database round-trips.

Otherwise this looks good.